### PR TITLE
Modified how dist_type is retrieved for 5 gsims which use self.dist_t…

### PR DESCRIPTION
**Pull request by CB-quakemodel**
…ype = REQUIRED_DISTANCES to make sure all gsims are compatible with the smtk comparison module (all distance types are assigned to each gmpe in the comparison utilities to ensure we can trellis plot of gm value against any core distance metric (even when the gmpe does not support the r_type, because we can still get the distance within the contexts and use it within plots), for these 5 gsims adding the core distance types meant the required distance type could not be assigned. A new function is added to these 5 gsims called _get_dist_type which ensures in each gsim the required distance metric is retrieved for each subclass. All tests still pass for the gsims, this just ensures they can also be used in the smt's comparison module.

### CODE REVIEW

File: openquake/hazardlib/gsim/bindi_2014.py [hunk 1]
```diff
@@ -114,6 +114,34 @@ def _get_style_of_faulting_term(C, ctx):
     return C["sofN"] * NS + C["sofR"] * RS + C["sofS"] * SS
 
 
+def _get_dist_type(GMPE):
+    """
+    Get distance type required for the corresponding class of 
+    Bindi et al. (2014)
+    """
+    # If gsim is [BindiEtAl2014RjbArmenia] strip to [BindiEtAl2014Rjb]
+    toml_str = GMPE._toml.split('\n')[0]
```
Comment (micheles) [inline openquake/hazardlib/gsim/bindi_2014.py:123]: This is horrible. Why do you need to access a private attribute? Why not simply using GMPE.__class__.__name__? Also all caps are reserved to constants, you should use `gmpe` as variable name.

File: openquake/hazardlib/gsim/bindi_2014.py [hunk 2]
```diff
@@ -114,6 +114,34 @@ def _get_style_of_faulting_term(C, ctx):
...
+                    str(BindiEtAl2014RjbEC8()), 
+                    str(BindiEtAl2014RjbEC8NoSOF())]
+    
+    for model in rjb_variants:
+        if model == toml_str or model == gmpe_str:
+            dist_type = {'rjb'}
+    
+    rhypo_variants = [str(BindiEtAl2014Rhyp()),
+                      str(BindiEtAl2014RhypEC8()),
+                      str(BindiEtAl2014RhypEC8NoSOF())]
```
Comment (micheles) [inline openquake/hazardlib/gsim/bindi_2014.py:136]: Using `str` looks like a code smell. Do you want the name of the class? You can use `__name__` in that case.

File: openquake/hazardlib/gsim/lanzano_2016.py [hunk 3]
```diff
@@ -151,6 +151,23 @@ class C: Vs30 < 360 m/s
...
 
 
+def _get_dist_type(gmpe):
+    """
+    Get distance type required for the corresponding class of 
+    Lanzano et al. 2016
+    """
+    # Get gsim class name
+    gsim = gmpe.__class__().__class__.__name__
+
```
Comment (micheles) [inline openquake/hazardlib/gsim/lanzano_2016.py:161]: This is absurd. Just use  `gmpe.__class__.__name__`

File: openquake/hazardlib/gsim/bindi_2014.py [hunk 4]
```diff
@@ -114,6 +114,27 @@ def _get_style_of_faulting_term(C, ctx):
...
+                      BindiEtAl2014RhypEC8NoSOF.__name__]
+    
+    # Assign required distance type
+    if gsim in rhypo_variants:
+        dist_type = 'rhypo'
+    else:
+        dist_type = 'rjb'
+    
+    return [dist_type]
+
```
Comment (micheles) [inline openquake/hazardlib/gsim/bindi_2014.py:136]: Again, it is not clear why you are returning a list with a single argument and not a scalar.

File: openquake/hazardlib/gsim/bindi_2014.py [hunk 5]
```diff
@@ -114,6 +114,27 @@ def _get_style_of_faulting_term(C, ctx):
     return C["sofN"] * NS + C["sofR"] * RS + C["sofS"] * SS
 
 
+def _get_dist_type(gmpe):
+    """
+    Get distance type required for the corresponding class of 
+    Bindi et al. (2014)
+    """
+    # Get gsim class name
```
Comment (micheles) [inline openquake/hazardlib/gsim/bindi_2014.py:122]: This is ugly. Why not
```python
if  "Rhyp" in gmpe.__class__.__name__
```
?

File: openquake/hazardlib/gsim/bindi_2017.py [hunk 6]
```diff
@@ -80,6 +80,23 @@ def _get_site_term(C, vs30):
...
 
 
+def _get_dist_type(gmpe):
+    """
+    Get distance type required for the corresponding class of 
+    Bindi et al. (2017)
+    """
+    # Get gsim class name
+    gsim = gmpe.__class__().__class__.__name__
+
```
Comment (micheles) [inline openquake/hazardlib/gsim/bindi_2017.py:90]: As above

File: openquake/hazardlib/gsim/drouet_alpes_2015.py [hunk 7]
```diff
@@ -669,3 +669,42 @@ class DrouetAlpes2015RrupHR_50bars(DrouetAlpes2015Rrup):
...
+    
+    # Get required dist type
+    if gsim in repi_variants:
+        dist_type = 'repi'
+    elif gsim in rjb_variants:
+        dist_type = 'rjb'
+    elif gsim in rrup_variants:
+        dist_type = 'rrup'
+    elif gsim in rhypo_variants:
+        dist_type = 'rhypo'
```
Comment (micheles) [inline openquake/hazardlib/gsim/drouet_alpes_2015.py:708]: Even this can be simplified as above

File: openquake/hazardlib/gsim/lanzano_2019.py [hunk 8]
```diff
@@ -95,6 +95,27 @@ def _get_mechanism(ctx, C):
...
+    rjb_variants = [LanzanoEtAl2019_RJB_OMO.__name__,
+                    LanzanoEtAl2019_RJB_OMOscaled.__name__,
+                    LanzanoEtAl2019_RJB_OMO_RefRock.__name__]
+
+    # Get required distance type
+    if gsim == LanzanoEtAl2019_RUP_OMO.__name__:
+        dist_type = 'rrup'
+    elif gsim in rjb_variants:
+        dist_type = 'rjb'
+
```
Comment (micheles) [inline openquake/hazardlib/gsim/lanzano_2019.py:115]: Can we ever reach the missing clause?

**Comment by CB-quakemodel** [comment]
> Relying on a convention about the class name is still not ideal. Since we have the mandatory attribute .REQUIRES_DISTANCES, why not using something like
> 
> ```python
> dist_type = 'rhypo' if 'rhypo' in gmpe.REQUIRES_DISTANCES else 'rjb'
> ```
> 
> ? You do not need to define an helper function, since it is just one line of code.

The modification is to ensure these 5 gsims work in the SMT where I'm adding all of the 'core' distance metrics of epicentral, joyner-boore, rupture and hypocentral distances to GMPE.REQUIRES_DISTANCES so all of them can be extracted from the context for plotting of ground-motion intensities vs distance. If GMPE.REQUIRES_DISTANCES is used the wrong class will be assigned, so I think relying on the class name is the best (albeit not ideal) solution to this for now

**Comment by micheles** [comment]
I see, then LGTM.

**Review by micheles — APPROVED** [review]

**Pull request merged**